### PR TITLE
[bugfix] fix two edge cases in functionalization

### PR DESCRIPTION
**Pull request by bdhirsh**
This should fix two bugs I encountered while playing with conditional functionalization in functorch:
- correct handling for scalar-tensor inputs. They should be ignored by the functionalization pass, since you can't mutate/take out views on a scalar in python-land.
- Correct handling for mutations to the "original" tensor that a view was taken from: internally, they have an empty vector of `ViewMeta` objects, which requires special handling (they were hitting an underflow issue)

I added tests for both cases.

Stack from [ghstack](https://github.com/ezyang/ghstack):
* #68339 [bugfix] functionalization pass for view ops without a 'self' first argumennt
* **#68269 [bugfix] fix two edge cases in functionalization**

Differential Revision: [D32396357](https://our.internmc.facebook.com/intern/diff/D32396357)

### CODE REVIEW

**Comment by bdhirsh** [comment]
@bdhirsh has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32396357).

**Comment by bdhirsh** [comment]
@bdhirsh has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32396357).

File: aten/src/ATen/FunctionalTensorWrapper.cpp [hunk 1]
```diff
@@ -249,6 +253,10 @@ TensorList to_functional_tensor(const TensorList& t_list) {
 }
 
 Tensor from_functional_tensor(const Tensor& tensor) {
+  // Note [Wrapped Numbers <> Functionalization]
+  if (tensor.unsafeGetTensorImpl()->is_wrapped_number()) {
+      return tensor;
+  }
```
Comment (zou3519) [inline aten/src/ATen/FunctionalTensorWrapper.cpp:259]: What happens if one performs operations between FunctionalTensorWrapper and non-FunctionalTensorWrapper?

Comment (zou3519) [reply aten/src/ATen/FunctionalTensorWrapper.cpp:259]: Answered my question: if things go through the fallback, then the fallback "unwraps everything that is not a wrapped number" via from_functional_tensor, calls the op, and then re-wraps with to_functional_tensor. The generated code for in-place operations and views assumes that wrapped numbers don't occur

**Review by zou3519 — APPROVED** [review]
one question otherwise lgtm

File: aten/src/ATen/FunctionalTensorWrapper.cpp (hunk 1 above)
Comment (bdhirsh) [reply aten/src/ATen/FunctionalTensorWrapper.cpp:259]: yep, that's right. For functorch, this also means that mixed functionalized/non-functionalized tensors currently don't work; something like:
```
y = torch.ones(2)
def f(x):
   return x + y
out = functionalize(f)(torch.ones(2))
```

File: aten/src/ATen/FunctionalStorageImpl.cpp [hunk 2]
```diff
@@ -55,6 +55,8 @@ void Alias::add_update(const at::Tensor& updated_val, const std::vector<ViewMeta
 const Tensor apply_update(const Alias::Update& update, const Tensor& base) {
   at::Tensor t = update.new_val;
   TORCH_INTERNAL_ASSERT(!at::functionalization::impl::isFunctionalTensor(t));
+  if (update.view_metas.size() == 0) return t;
```
[the remark this answers is not in the document]
Comment (bdhirsh) [reply aten/src/ATen/FunctionalStorageImpl.cpp:58]: That case actually results in unsigned integer underflow - I forget exactly how I triggered it, but I encountered it during the conditional functionalization PR. `size()` returns an unsigned int, so `update.view_metas.size() - 1` underflows to the largest unsigned integer and we get UB.

That's why I added the extra test [here](https://github.com/pytorch/pytorch/pull/68269/files#diff-e302079b2ae45f6ada1a3cf98a49dfe5b518efbb51d2299de9d494246d4f74cbR73), which should trigger that case: you take out a view on a tensor, but then try to modify the original tensor (at which point it has an empty vector of `ViewMeta` objects).

[the remark this answers is not in the document]
Comment (zou3519) [reply aten/src/ATen/FunctionalStorageImpl.cpp:58]: Oh that makes sense, thanks for the clarification. I forgot about unsigned ints in C++...

[the remark this answers is not in the document]
Comment (bdhirsh) [reply aten/src/ATen/FunctionalStorageImpl.cpp:58]: me too 😛

[the remark this answers is not in the document]
Comment (ezyang) [reply aten/src/ATen/FunctionalStorageImpl.cpp:58]: Friends don't let friends use unsigned ints in C/C++ 😇

**Pull request closed without merging**